WIP: initial implementation of service/watcher server

jmuk opened the pull request.

ayj commented:
Thanks for getting this started. 

FYI, the plan of record is to use etcd for the reference storage backend (see https://github.com/istio/pilot/issues/850#issuecomment-308492572 and https://docs.google.com/document/d/13UqmYpEjNGT2xucIgWaTQIQUgQuAASeIdm8Ks3zp5Fk/edit#bookmark=id.de379zo7zqdl). We could keep the redis and fstore versions around but it might become too much work to improve and maintain something we don't ever use. It might be good to pull in the work that @ijsnellf already started in in https://github.com/istio/pilot/pull/870.

ayj commented on cmd/gals/cmd/server.go:
```diff
@@ -15,14 +15,19 @@
-	port uint16
+	storeURL    string
+	port        uint16
+	gatewayPort uint16
```
Maybe rename these and the corresponding flags `grpcPort` and `httpPort` to be more explicit?

jmuk replied:
Renamed -- but I think `restPort` might be better (y'know, technically grpc is on top of http/2)

ayj commented on cmd/gals/cmd/server.go:
```diff
@@ -38,15 +43,20 @@ func serverCmd(printf, fatalf shared.FormatFn) *cobra.Command {
 	serverCmd.PersistentFlags().Uint16Var(&sa.port, "port", 9096, "Galley API port")
+	serverCmd.PersistentFlags().Uint16Var(&sa.gatewayPort, "gateway-port", 9097, "the port for grpc-gateway to the API")
+	serverCmd.PersistentFlags().StringVar(&sa.storeURL, "store-url", "", "the URL for the backend storage")
+	serverCmd.PersistentFlags().UintVar(&sa.interval, "interval", 5000, "the interval to push updates in msecs")
```
nit: Use https://godoc.org/github.com/spf13/pflag#DurationVar for native golang `time.Duration` flag handling.

jmuk replied:
Good to know, thanks!  Based on a redesign, this interval flag has been removed for the time being.

ayj commented on pkg/server/galley.go:
```diff
@@ -15,17 +15,101 @@
+		return err
+	}
+
+	go http.ListenAndServe(fmt.Sprintf(":%d", gatewayPort), mux)
```
Can you move goroutine creation to the caller (i.e. Server.Start) so it can properly log errors? e.g.

jmuk replied:
done

ayj commented on pkg/server/service.go:
```diff
@@ -0,0 +1,119 @@
+	kvs store.KeyValueStore
+}
+
+var _ configpb.ServiceServer = &configAPIServer{}
```
Why the empty import here and elsewhere?

mandarjog replied:
to force compile time check of the interface.

jmuk replied:
right, for force checking -- but removed. server go handles the registration (`RegisterServiceServer`) can verify the same thing, so this checking is not necessary anymore.

ayj commented on pkg/server/service.go:
```diff
@@ -0,0 +1,119 @@
+	"istio.io/galley/pkg/store"
+)
+
+type configAPIServer struct {
```
s/configAPIServer/GalleyService ?

Also, can this be an exported type since `NewServiceServer` is exported?

jmuk replied:
Done

mandarjog commented:
I am updating the KV Store abstraction to use a channel based watcher to better align with etcdv3.

jmuk commented:
I'm working on etcd implementation too. shouldn't be big.

jmuk commented:
(renamed the title as WIP, since the presubmit should fail due to the lack of the proto files -- config/v1/service.proto is still on PR)

ijsnellf commented on pkg/server/galley.go:
```diff
@@ -15,17 +15,101 @@
+		grpc.WithCompressor(grpc.NewGZIPCompressor()),
+		grpc.WithDecompressor(grpc.NewGZIPDecompressor()),
+	}
+	err := configpb.RegisterServiceHandlerFromEndpoint(ctx, mux, fmt.Sprintf("localhost:%d", port), opts)
```
Shouldn't this be `0.0.0.0` instead of `localhost`?

jmuk replied:
This is a bit of confusing -- but actually localhost is correct. This is registering a grpc gateway, and specifying the target of the grpc service (which resides in the same process). renamed the variable names, I hope things get clearer.

ijsnellf commented on pkg/server/galley.go:
```diff
@@ -15,17 +15,101 @@
+	var listener net.Listener
+	var err error
+	// get the network stuff setup
+	if listener, err = net.Listen("tcp", fmt.Sprintf(":%d", port)); err != nil {
```
No need for `if ...; err != nil { ... }` idiom:
```go
listener, err = net.Listen("tcp", fmt.Sprintf(":%d", port))
if err != nil {
    ...
}
```

ijsnellf replied:
Listen on `0.0.0.0`?

jmuk replied:
former -- done. I think `0.0.0.0` is unnecessary, according to https://golang.org/pkg/net/#Listen, simply starting with `:` means same.

jmuk commented:
etcd support is in (the redis backend is missing watcher feature now)

rshriram commented on pkg/store/inventory/BUILD:
```diff
@@ -0,0 +1,15 @@
+        "//pkg/store:go_default_library",
+        "//pkg/store/etcd:go_default_library",
+        "//pkg/store/redis:go_default_library",
+    ],
```
Why do we need redis here?

rshriram commented on pkg/store/redis/BUILD:
```diff
@@ -0,0 +1,29 @@
+package(default_visibility = ["//visibility:public"])
+
```
We had agreed that we were not going to build a redis backend for configuration. 
So, what would be the purpose of this redis store? 
What is the HA story for Redis configuration store and the consistency semantics? Redis cluster does not have a strong consistency model. So what happens if ACLs or rules are lost? 

cc @smawson @louiscryan

jmuk commented:
Removed the redis part, and reformed everything on top of @mandarjog's https://github.com/istio/galley/pull/37. I'll split the etcd implementation into another PR when #37 is merged.

mandarjog commented on pkg/server/watcher.go:
```diff
@@ -0,0 +1,226 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	// watch the whole updates.
+	c, err := s.kvs.Watch(ctx, "/", 0)
+	if err != nil {
```
if error then log ?

jmuk replied:
added the log line.

mandarjog commented on pkg/server/watcher.go:
```diff
@@ -0,0 +1,226 @@
+func (s *WatcherServer) pushEvents(evs []store.Event) {
+	s.mu.Lock()
+	toSend := map[int64]*galleypb.WatchEvents{}
+	for _, ev := range evs {
```
suggestion: can we split that up into 2 functions ?

1st loop = filterEvents()
2nd loop = sendEvents()

jmuk replied:
done

mandarjog commented on pkg/store/etcd/etcdstore.go:
```diff
@@ -0,0 +1,155 @@
+
+// Watch implements a store.Watcher method.
+func (es *KeyValue) Watch(ctx context.Context, key string, revision int64) (<-chan store.Event, error) {
+	c := make(chan store.Event)
```
should be a buffered channel ?

jmuk replied:
This file ended up in another PR. Done there -- see #38

ayj commented on pkg/server/watcher.go:
```diff
@@ -0,0 +1,226 @@
+	cancelWatchChan context.CancelFunc
+	nextWatcherID   int64
+	watchers        map[int64]*watcher
+	mu              sync.Mutex
```
What field(s) does `mu` protect? Standard convention seems to be `mu` preceding the protected variables, e.g.

```golang
unprotectedField int

mu              sync.Mutex
protectedField0 int
protectedField1 int
protectedField2 int


```

jmuk replied:
reordered the field in the style.

ayj commented on pkg/server/util.go:
```diff
@@ -0,0 +1,80 @@
+	}, nil
+}
+
+func buildObject(data []byte, revision int64) (obj *galleypb.ConfigObject, err error) {
```
Named return parameters don't add anything here to readability or conciseness. Maybe better to write as the following?

```golang
func buildObject(data []byte, revision int64) (*galleypb.ConfigObject, error) {
    var obj galleypb.ConfigObject
    if err := proto.Unmarshal(data, &obj); err != nil {
        return nil, err
    }
    obj.Meta.Revision = revision
    return &obj, nil
}


```

jmuk replied:
done

ayj commented on pkg/server/util.go:
```diff
@@ -0,0 +1,80 @@
+	return obj, nil
+}
+
+func readKvsToObjects(kvs store.KeyValue, prefix string) (objs []*galleypb.ConfigObject, revision int64, err error) {
```
Similar comment to line 53 re: named return parameters. Further more, the named `revision` and `err` are shadowed by line 64.

jmuk replied:
done

ayj commented on pkg/server/watcher.go:
```diff
@@ -0,0 +1,226 @@
+	}
+}
+
+func (s *WatcherServer) pushEvents(evs []store.Event) {
```
Maybe add a comment that this serializes dispatching events across all watchers? This is probably okay for initial version but any concerns about performance w.r.t larger number of events/watchers and blocking `stream.Send()`?

jmuk replied:
Added some comments here.  Regarding the performance -- I actually didn't think of that situation deeply. Added comment for that too.

ayj commented on pkg/server/watcher.go:
```diff
@@ -0,0 +1,226 @@
+		case <-t:
+			s.pushEvents(evs)
+			evs = []store.Event{}
+		}
```
1) This loop needs to exit when the context is canceled, e.g.

```golang
case <-ctx.Done():
    return
```

2) Even with change #1 above, I don't think it is guaranteed that this loop will exit before a new instance of the loop is started. It looks like this would result in a newly started watcher receiving pushed events sooner than the specified watch interval, so maybe this isn't a problem though?

jmuk replied:
1. you're totally right. added. Thanks!
2. I think that's no problem. The two ongoing goroutines wouldn't conflict each other here.

kyessenov commented:
It's very difficult to review 1.5k PR. Maybe split API definition out so we can discuss that?

rshriram commented:
Split API code and bazel ? We are not st a stage where e2e has to pass.
On Mon, Jun 26, 2017 at 5:34 PM Kuat <notifications@github.com> wrote:

> It's very difficult to review 1.5k PR. Maybe split API definition out so
> we can discuss that?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/galley/pull/28#issuecomment-311188715>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AH0qd1fBK1KXTu9dq-mXfUvTeIpuj92Gks5sICPcgaJpZM4N7wDq>
> .
>
-- 

~shriram

jmuk commented:
I split this PR into 3 parts according to your comments -- #38 for pkg/store, and #39 for bazel things. Now this PR only contains the gRPC server implementation only.  I haven't addressed the review comments above, will do soon.

jmuk commented:
sadly 😞, splitting bazel didn't help for the presubmit status -- linters anyways check the go files and fail because they lack dependencies, especially pb.go files (i.e. https://github.com/istio/api/pull/122).

jmuk commented:
closing. Galley service is already merged. Watcher will be reimplemented with the new proto definition.
